src/Raven.Server/Documents/Expiration/ExpirationStorage.cs

```diff
@@ -146,38 +159,90 @@ private void ThrowWrongExpirationDateFormat(Slice lowerId, string expirationDate
             }
         }
 
+        private (bool AllExpired, LazyStringValue Id) GetConflictedExpiration(DocumentsOperationContext context, DateTime currentTime, Slice clonedId)
+        {
+            LazyStringValue id = null;
+            var allExpired = true;
+            var conflicts = _database.DocumentsStorage.ConflictsStorage.GetConflictsFor(context, clonedId);
+            if (conflicts.Count > 0)
+            {
+                foreach (var conflict in conflicts)
+                {
+                    id = conflict.Id;
+
+                    if (HasExpired(conflict.Doc, currentTime))
+                        continue;
+
+                    allExpired = false;
+                    break;
+                }
+            }
+
+            return (allExpired, id);
+        }
+
         public static bool HasExpired(BlittableJsonReaderObject data, DateTime currentTime)
         {
             // Validate that the expiration value in metadata is still the same.
             // We have to check this as the user can update this value.
-            if (data.TryGet(Constants.Documents.Metadata.Key, out BlittableJsonReaderObject metadata) == false ||
-                metadata.TryGet(Constants.Documents.Metadata.Expires, out string expirationDate) == false)
+            if (data.TryGet(Constants.Documents.Metadata.Key, out BlittableJsonReaderObject metadata) == false)
                 return false;
 
-            if (DateTime.TryParseExact(expirationDate, DefaultFormat.DateTimeFormatsToRead, CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind, out var date) == false)
-                return false;
-
-            if (currentTime < date.ToUniversalTime())
-                return false;
+            return HasExpired(metadata, Constants.Documents.Metadata.Expires, currentTime);
+        }
 
-            return true;
+        private static bool HasExpired(BlittableJsonReaderObject metadata, string metadataPropertyName, DateTime currentTime)
+        {
+            if (metadata.TryGet(metadataPropertyName, out string expirationDate))
+            {
+                if (DateTime.TryParseExact(expirationDate, DefaultFormat.DateTimeFormatsToRead, CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind, out var date))
+                {
+                    if (currentTime >= date.ToUniversalTime())
+                        return true;
+                }
+            }
+            return false;
         }
 
         public int DeleteExpiredDocuments(DocumentsOperationContext context, Dictionary<Slice, List<(Slice LowerId, LazyStringValue Id)>> expired)
```

Comment:
The name here is now confusing since it is now 'Trigger' some actions and not just delete documents

Reply:
+1